### PR TITLE
Map our user type to a table

### PR DESCRIPTION
**Pull request by SamGodwin2**
### What is the context of this PR?
This Pr adds the concept of a user table and is part of the inroductionary work for permissions.

### How to review 
New user table is correctly created and on first sign in user accounts are created for all users. Tests also pass.

### CODE REVIEW

File: eq-author-api/db/models/DynamoDB.js [hunk 1]
```diff
@@ -79,6 +84,33 @@ const questionnaireVersionsSchema = new dynamoose.Schema(
...
+    sub: {
+      type: String,
+      required: true,
+    },
+    picture: {
+      type: String,
+    },
+  },
+  {
+    throughput: throughput,
```
Comment (tgandrews) [inline eq-author-api/db/models/DynamoDB.js]:
```suggestion
    throughput,
```

File: eq-author-api/middleware/addUserToRequest.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const { getUserBySub } = require("../utils/datastore");
+
+module.exports = async (req, res, next) => {
+  let user = await getUserBySub(req.auth.sub);
+
+  if (!user) {
+    res.status(401).send("User does not exist");
```
Comment (tgandrews) [inline eq-author-api/middleware/addUserToRequest.js]: The rest of the middleware will keep running. So you should return here.

File: eq-author-api/middleware/upsertUser.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+module.exports = async (req, res) => {
+  let user = await getUserBySub(req.auth.sub);
+
+  if (!user) {
+    user = await createUser(req.auth);
+  } else if (!checkForUpdates(user, req.auth)) {
+    Object.assign(user, req.auth);
+    user = await updateUser(user);
+  }
+  res.send(200);
```
Comment (tgandrews) [inline eq-author-api/middleware/upsertUser.js]: You should send a json response here:
`res.json({ status: "OK" });`
It by default will have a status of 200.

File: eq-author-api/middleware/upsertUser.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  return user;
```
Comment (tgandrews) [inline eq-author-api/middleware/upsertUser.js]: I don't think we should be returning this.
```suggestion
  return;
```

File: eq-author-api/middleware/upsertUser.test.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  });
+  it("should create a user is one doesn't exist", async () => {
+    const user = await upsertUser(req, res);
+    expect(user).toMatchObject({
+      sub: "new_user",
+      id: expect.any(String),
+    });
+    expect(res.send).toHaveBeenCalledWith(200);
+  });
+  it("should update the user if any changes", async () => {
```
Comment (tgandrews) [inline eq-author-api/middleware/upsertUser.test.js]: We are also missing a test for when the user is already created and not changed.

File: eq-author-api/schema/resolvers/base.js [hunk 6]
```diff
@@ -219,17 +219,17 @@ const Resolvers = {
       return { pageId, ...confirmationPage };
     },
     me: (root, args, ctx) => ({
-      id: ctx.auth.sub,
-      ...pick(ctx.auth, ["name", "email", "picture"]),
-      name: ctx.auth.name || ctx.auth.email,
+      id: ctx.user.id,
+      ...pick(ctx.user, ["name", "email", "picture"]),
```
Comment (tgandrews) [inline eq-author-api/schema/resolvers/base.js]: We pick name and then overwrite it?

File: eq-author-api/schema/resolvers/base.js [hunk 7]
```diff
@@ -219,17 +219,17 @@ const Resolvers = {
...
+      name: ctx.user.name || ctx.user.email,
```
Comment (tgandrews) [inline eq-author-api/schema/resolvers/base.js]: Should this be `displayName`?

File: eq-author-api/utils/datastoreDynamo.js [hunk 8]
```diff
@@ -21,6 +23,41 @@ const saveModel = (model, options = {}) =>
     });
   });
 
+const createUser = user => {
+  const { id, email, name, sub, picture } = user;
+  return saveModel(
+    new UserModel({
+      id: id ? id : uuid.v4(),
```
Comment (tgandrews) [inline eq-author-api/utils/datastoreDynamo.js]: Sometimes it is created with a known id?

```suggestion
      id: id || uuid.v4(),
```

File: eq-author-api/utils/datastoreFileSystem.js [hunk 9]
```diff
@@ -1,5 +1,6 @@
-const { omit, reject } = require("lodash");
+const { omit, reject, merge } = require("lodash");
```
Comment (tgandrews) [inline eq-author-api/utils/datastoreFileSystem.js]: @samiwel are we going to keep this around or should we just drop this filesystem support. It seems an unnecessary burden to maintain.

File: eq-author/.env.development [hunk 10]
```diff
@@ -1,2 +1,3 @@
 REACT_APP_API_URL="http://localhost:4000/graphql"
```
Comment (tgandrews) [inline eq-author/.env.development]: We can drop the server location for all of these now?

File: eq-author/src/App/SignInForm/index.js [hunk 11]
```diff
@@ -43,7 +43,13 @@ SignInForm.defaultProps = {
...
     credentialHelper,
     callbacks: {
-      signInSuccessWithAuthResult: () => false, // Avoid redirects after sign-in.
+      signInSuccessWithAuthResult: authResult => {
+        return window
+          .fetch(process.env.REACT_APP_SIGN_IN_URL, {
+            headers: { authorization: `Bearer ${authResult.user.ra}` },
+          })
+          .then(() => false);
+      }, // Avoid redirects after sign-in.
```
Comment (tgandrews) [inline eq-author/src/App/SignInForm/index.js]: Is this `false` and comment still necessary?

**Comment by SamGodwin2** [comment]
[Map our User type to a table. (6)](https://trello.com/c/b5tDb8j3/976-map-our-user-type-to-a-table-6)

File: eq-author-api/utils/datastoreFileSystem.js (hunk 9 above)
Comment (samiwel) [reply eq-author-api/utils/datastoreFileSystem.js]: Now that we have some kind of GUI for local dynamodb and an import script to seed the tables, maintaining a filesystem datastore is not as important.

File: eq-author-api/app.js [hunk 12]
```diff
@@ -45,14 +49,17 @@ app.use(
   pino,
   cors(),
   createAuthMiddleware(logger),
-  loadQuestionnaire,
-  runQuestionnaireMigrations(logger)(require("./migrations"))
+  runQuestionnaireMigrations(logger)(require("./migrations")),
+  userSetup,
```
Comment (tgandrews) [inline eq-author-api/app.js]: Should this also be set for launch? We could do as part of the permissions changes.

**Pull request closed without merging**